app/sidero-controller-manager/api/v1alpha1/server_types.go

```diff
@@ -125,36 +175,57 @@ func PartialEqual(a, b interface{}) bool {
 		new = new.Elem()
 	}
 
-	for i := 0; i < old.NumField(); i++ {
-		if old.Field(i).IsZero() {
-			// Skip zero values, since that indicates that the user did not supply
-			// the field, and does not want to compare it.
-			continue
-		}
+	// Skip invalid or zero values, since that indicates that the user
+	// did not supply the field, and does not want to compare it.
+	if !old.IsValid() || old.IsZero() {
+		return true
+	}
 
-		f1 := old.Field(i).Interface()
-		f2 := new.Field(i).Interface()
+	if old.Kind() == reflect.Struct && new.Kind() == reflect.Struct {
+		// Recursively compare structs
+		for i := 0; i < old.NumField(); i++ {
+			f1 := old.Field(i).Interface()
+			f2 := new.Field(i).Interface()
 
-		if f1 != f2 {
+			if !PartialEqual(f1, f2) {
+				return false
+			}
+		}
+	} else if old.Kind() == reflect.Slice && new.Kind() == reflect.Slice {
+		// Skip slices where the requested slice is larger than the actual slice,
+		// as that indicates that the user wants to filter for more
+		// processors/memory modules/storage devices than are present.
+		if old.Len() > new.Len() {
 			return false
 		}
+		// Recursively compare slices
+		for i := 0; i < old.Len(); i++ {
+			f1 := old.Index(i).Interface()
+			f2 := new.Index(i).Interface()
+
+			if !PartialEqual(f1, f2) {
+				return false
+			}
+		}
+	} else {
+		// Directly compare values, but only if the actual value is valid.
+		return new.IsValid() && old.Interface() == new.Interface()
 	}
 
 	return true
 }
 
 // ServerSpec defines the desired state of Server.
 type ServerSpec struct {
-	EnvironmentRef    *corev1.ObjectReference `json:"environmentRef,omitempty"`
-	Hostname          string                  `json:"hostname,omitempty"`
-	SystemInformation *SystemInformation      `json:"system,omitempty"`
-	CPU               *CPUInformation         `json:"cpu,omitempty"`
-	BMC               *BMC                    `json:"bmc,omitempty"`
-	ManagementAPI     *ManagementAPI          `json:"managementApi,omitempty"`
-	ConfigPatches     []ConfigPatches         `json:"configPatches,omitempty"`
-	Accepted          bool                    `json:"accepted"`
-	Cordoned          bool                    `json:"cordoned,omitempty"`
-	PXEBootAlways     bool                    `json:"pxeBootAlways,omitempty"`
+	EnvironmentRef *corev1.ObjectReference `json:"environmentRef,omitempty"`
+	Hardware       *HardwareInformation    `json:"hardware,omitempty"`
```

Comment:
this is a breaking change for the CRDs, as the fields were removed. We could do also migration and new CRD version, but that might need more changes. I wonder if we could still keep things same way in the CRDs, even if we restructure the fields in Go structs.

Reply:
Yeah, that is also what I was struggling with. I agree not breaking things is better, but I honestly don't know a good way.
A partial migration should be possible I think, although it will probably always lead to incomplete information (for example servers with multiple CPU's will only have 1 listed). Also we would only have a few fields of the new set available.
I'm not sure what is worse: not having information or having incomplete information...

Reply:
:+1: for a new CRD version and we do CRD migrations, it'll be less visible changes for a user

Reply:
I agree we should probably do new API version and migration, but this means we should also fix the webhooks for this controller (it's right now broken/missing) and go all the way with proper migration. And if we do that, should we do more changes as well?